core/services/llo/channeldefinitions/onchain_channel_definition_cache.go

```diff
@@ -425,104 +768,157 @@ func (c *channelDefinitionCache) fetchChannelDefinitions(ctx context.Context, ur
 	hash := sha3.New256()
 	// Stream the data directly into the hash and copy to buf as we go
 	if _, err := io.Copy(hash, teeReader); err != nil {
-		return nil, fmt.Errorf("failed to read from body: %w", err)
+		return nil, fmt.Errorf("failed to read channel definitions response body from %s: %w", trigger.URL, err)
 	}
 
 	actualSha := hash.Sum(nil)
-	if !bytes.Equal(expectedSha[:], actualSha) {
-		return nil, fmt.Errorf("SHA3 mismatch: expected %x, got %x", expectedSha, actualSha)
+	if !bytes.Equal(trigger.SHA[:], actualSha) {
+		return nil, fmt.Errorf("SHA3 mismatch for channel definitions from %s: expected %x, got %x", trigger.URL, trigger.SHA, actualSha)
 	}
 
 	var cd llotypes.ChannelDefinitions
 	decoder := json.NewDecoder(&buf)
 	if err := decoder.Decode(&cd); err != nil {
-		return nil, fmt.Errorf("failed to decode JSON: %w", err)
+		return nil, fmt.Errorf("failed to decode channel definitions JSON from %s: %w", trigger.URL, err)
+	}
+
+	// Annotate each definition with its source identifier.
+	for channelID, def := range cd {
+		def.Source = trigger.Source
+		cd[channelID] = def
 	}
 
 	return cd, nil
 }
 
-////////////////////////////////////////////////////////////////////
-// Persistence
-////////////////////////////////////////////////////////////////////
-
-func (c *channelDefinitionCache) persist(ctx context.Context) (memoryVersion, persistedVersion uint32, err error) {
-	c.persistMu.RLock()
-	persistedVersion = c.persistedVersion
-	c.persistMu.RUnlock()
+// persist atomically writes the in-memory source definitions (c.definitions.Sources) to the database.
+// Returns the memory and persisted block numbers along with any error that occurred during persistence.
+func (c *channelDefinitionCache) persist(ctx context.Context) (memoryBlockNum, persistedBlockNum int64, err error) {
```

Comment:
I would remove the explicit return parameter. See the problem below
```suggestion
func (c *channelDefinitionCache) persist(ctx context.Context) (int64, int64, error) {
```